File: VB/DxSample.Service/Models/DataContext.vb
Imports System.Collections.Generic
Imports System.Linq
Imports System.Threading.Tasks

Namespace DxSample.Service.Models
	Public NotInheritable Class DataContext

		Private Sub New()
		End Sub

		Private Shared ListCustomers As List(Of Customer)
		Private Shared ListOrders As List(Of Order)
		Private Shared CustomersCounter As Integer
        Private Shared OrdersCounter As Integer
        Private Shared ReadOnly Sync As New Object

		Shared Sub New()
			DataContext.ListCustomers = New List(Of Customer)()
			DataContext.ListOrders = New List(Of Order)()
			DataContext.CustomersCounter = 1
			DataContext.OrdersCounter = 1
			DataContext.AddCustomer("John")
			DataContext.AddCustomer("Bob")
			DataContext.AddOrder("Chai", 1)
			DataContext.AddOrder("Chang", 1)
			DataContext.AddOrder("Queso Caprale", 2)
		End Sub

		Public Shared ReadOnly Property Customers() As IQueryable(Of Customer)
			Get
				Return DataContext.ListCustomers.AsQueryable()
			End Get
		End Property

		Public Shared ReadOnly Property Orders() As IQueryable(Of Order)
			Get
				Return DataContext.ListOrders.AsQueryable()
			End Get
		End Property

		Public Shared Function AddCustomer(ByVal name As String) As Task(Of Customer)
            Return Task.Run(Of Customer)(Function()
                                             SyncLock DataContext.Sync
                                                 Dim result As New Customer() With {
                                                     .ID = DataContext.CustomersCounter,
                                                     .Name = name}
                                                 DataContext.CustomersCounter = DataContext.CustomersCounter + 1
                                                 DataContext.ListCustomers.Add(result)
                                                 Return result
                                             End SyncLock

                                         End Function)
		End Function

		Public Shared Function AddOrder(ByVal name As String, ByVal customerID As Integer) As Task(Of Order)
            Return Task.Run(Of Order)(Function()
                                          SyncLock DataContext.Sync
                                              Dim result As New Order() With {
                                                  .ID = DataContext.OrdersCounter,
                                                  .Name = name,
                                                  .CustomerID = customerID}
                                              DataContext.OrdersCounter = DataContext.OrdersCounter + 1
                                              DataContext.ListOrders.Add(result)
                                              Return result
                                          End SyncLock
                                      End Function)
		End Function

		Public Shared Function FindCustomerAsync(ByVal key As Integer) As Task(Of Customer)
			Return Task.Run(Of Customer)(Function() DataContext.Customers.SingleOrDefault(Function(c) c.ID = key))
		End Function

		Public Shared Function FindOrderAsync(ByVal key As Integer) As Task(Of Order)
			Return Task.Run(Of Order)(Function() DataContext.Orders.SingleOrDefault(Function(o) o.ID = key))
		End Function

		Public Shared Function UpdateCustomer(ByVal customer As Customer) As Task
			Return Task.Run(Sub()
				Dim original As Customer = DataContext.Customers.Single(Function(c) c.ID = customer.ID)
				Dim index As Integer = DataContext.ListCustomers.IndexOf(original)
				DataContext.ListCustomers(index) = customer
			End Sub)
		End Function

		Public Shared Function UpdateOrder(ByVal order As Order) As Task
			Return Task.Run(Sub()
				Dim original As Order = DataContext.Orders.Single(Function(o) o.ID = order.ID)
				Dim index As Integer = DataContext.ListOrders.IndexOf(original)
				DataContext.ListOrders(index) = order
			End Sub)
		End Function

		Public Shared Function RemoveCustomer(ByVal customer As Customer) As Task
			Return Task.Run(Function() DataContext.ListCustomers.Remove(customer))
		End Function

		Public Shared Function RemoveOrder(ByVal order As Order) As Task
			Return Task.Run(Function() DataContext.ListOrders.Remove(order))
		End Function
	End Class
End Namespace

File: CS/DxSample.Service/Models/DataContext.cs
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;

namespace DxSample.Service.Models {
    public static class DataContext {
        private static readonly List<Customer> ListCustomers;
        private static readonly List<Order> ListOrders;
        private static int CustomersCounter;
        private static int OrdersCounter;
        private static readonly object Sync = new object();

        static DataContext() {
            DataContext.ListCustomers = new List<Customer>();
            DataContext.ListOrders = new List<Order>();
            DataContext.CustomersCounter = 1;
            DataContext.OrdersCounter = 1;
            DataContext.AddCustomer("John");
            DataContext.AddCustomer("Bob");
            DataContext.AddOrder("Chai", 1);
            DataContext.AddOrder("Chang", 1);
            DataContext.AddOrder("Queso Caprale", 2);
        }

        public static IQueryable<Customer> Customers {
            get { return DataContext.ListCustomers.AsQueryable(); }
        }

        public static IQueryable<Order> Orders {
            get { return DataContext.ListOrders.AsQueryable(); }
        }

        public static Task<Customer> AddCustomer(string name) {
            return Task.Run<Customer>(() => {
                lock (DataContext.Sync) {
                    Customer result = new Customer() { ID = DataContext.CustomersCounter++, Name = name };
                    DataContext.ListCustomers.Add(result);
                    return result;
                }
            });
        }

        public static Task<Order> AddOrder(string name, int customerID) {
            return Task.Run<Order>(() => {
                lock (DataContext.Sync) {
                    Order result = new Order() { ID = DataContext.OrdersCounter++, Name = name, CustomerID = customerID };
                    DataContext.ListOrders.Add(result);
                    return result;
                }
            });
        }

        public static Task<Customer> FindCustomerAsync(int key) {
            return Task.Run<Customer>(() => DataContext.Customers.SingleOrDefault(c => c.ID == key));
        }

        public static Task<Order> FindOrderAsync(int key) {
            return Task.Run<Order>(() => DataContext.Orders.SingleOrDefault(o => o.ID == key));
        }

        public static Task UpdateCustomer(Customer customer) {
            return Task.Run(() => {
                Customer original = DataContext.Customers.Single(c => c.ID == customer.ID);
                int index = DataContext.ListCustomers.IndexOf(original);
                DataContext.ListCustomers[index] = customer;
            });
        }

        public static Task UpdateOrder(Order order) {
            return Task.Run(() => {
                Order original = DataContext.Orders.Single(o => o.ID == order.ID);
                int index = DataContext.ListOrders.IndexOf(original);
                DataContext.ListOrders[index] = order;
            });
        }

        public static Task RemoveCustomer(Customer customer) {
            return Task.Run(() => DataContext.ListCustomers.Remove(customer));
        }

        public static Task RemoveOrder(Order order) {
            return Task.Run(() => DataContext.ListOrders.Remove(order));
        }
    }
}